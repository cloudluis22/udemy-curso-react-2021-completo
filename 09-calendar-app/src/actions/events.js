import { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } from "react-dom";
import Swal from "sweetalert2";
import { convertEvents } from "../helpers/convertEvents";
import { fetchToken } from "../helpers/fetch";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async ( dispatch, getState ) => {

        const { uid, name } = getState().auth;

        try {
            const dbEvent = {
                ...event,
                start: event.startDate,
                end: event.endDate
            }

            const response = await fetchToken('events', dbEvent, 'POST');
            const body = await response.json();

            if (body.ok) {
                event.id = body.event.id
                event.user = {
                    _id: uid,
                    name: name
                }
                console.log(event);
                dispatch( eventAddNew(event) );
            }

        } catch (error) {
            console.log(error)
        }

    };
};

export const eventStartLoading = () => {
    return async (dispatch) => {

        try {
            const response = await fetchToken('events');
            const body = await response.json();

            const events = convertEvents(body.events);
           dispatch( eventsLoaded( events ) );

        } catch (error) {

        }

    };
};

export const eventStartUpdate = (event) => {
    return async (dispatch) => {

        try {

            const response = await fetchToken(`events/${event.id}`, event, 'DELETE');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventUpdated(event) );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
};

export const eventStartDelete = () => {


    return async (dispatch, getState) => {
        try {

            const { id } = getState().calendar.activeEvent;

            const response = await fetchToken(`events/${id}`, {}, 'delete');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventDeleted() );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
}

export const eventLogoutCleaning = () => ({ type: types.eventLogoutCleaning });

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

const eventUpdated = ( event ) => ({
    type: types.eventUpdated,
    payload: event
});

const eventDeleted = () => ({
    type: types.eventDeleted
});

const eventsLoaded = (events) => ({ type: types.eventsLoaded, payload: events });