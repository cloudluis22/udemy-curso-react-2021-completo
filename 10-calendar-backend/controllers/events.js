const { response } = require('express');
const Event = require('../models/Event-model');

const getEvents = async (request, response = response) => {

    const events = await Event.find()
        .populate('user', 'name');

    response.json({
        ok: true,
        events
    });
};

const createEvent = async (request, response = response) => {

    const event = new Event(request.body);

    try {

        event.user = request.uid;

        const savedEvent = await event.save();

        response.json({
            ok: true,
            event: savedEvent
        });

    } catch (error) {
        console.log(error);
        response.status(500).json({
            ok: false,
            msg: 'Unexpected error, please contact the admin.'
        });
    };

};

const updateEvent = async (request, response = response) => {

    const eventId = request.params.id;
    const uid = request.uid;

    try {

        const event = await Event.findById(eventId);

        if (!event) {
           return response.status(404).json({
                ok: false,
                msg: 'Event not found with given ID.'
            });
        };

        if (event.user.toString() !== uid) {
            console.log(event.user.toString());
            return response.status(401).json({
                ok: false,
                msg: 'Not enough privileges to modifiy this event.'
            });
        };

        const newEvent = {
            ...request.body,
            user: uid
        };

        const updatedEvent = await Event.findByIdAndUpdate(eventId, newEvent, { new: true, useFindAndModify: false });

        response.json({
            ok: true,
            event: updatedEvent
        });

    } catch (error) {
        console.log(error);
        response.status(500).json({
            ok: false,
            msg: 'Unexpected error, please contact the admin.'
        });
    };

};

const deleteEvent = async (request, response = response) => {

    const eventId = request.params.id;
    const uid = request.uid;

    try {

        const event = await Event.findById(eventId);


        if (!event) {
           return response.status(404).json({
                ok: false,
                msg: 'Event not found with given ID.'
            });
        };

        if (event.user.toString() !== uid) {
            return response.status(401).json({
                ok: false,
                msg: 'Not enough privileges to delete this event.'
            });
        };

        await Event.findByIdAndDelete(eventId);

        response.json({
            ok: true,
            msg: 'Event deleted'
        });

    } catch (error) {
        console.log(error);
        response.status(500).json({
            ok: false,
            msg: 'Unexpected error, please contact the admin.'
        });
    }


};

module.exports = {
    getEvents,
    createEvent,
    updateEvent,
    deleteEvent
};