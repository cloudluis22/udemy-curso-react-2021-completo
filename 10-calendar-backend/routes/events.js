/*
    Event Routes
    /api/events
*/
const { Router } = require('express');
const { check } = require('express-validator');

const { isDate } = require('../helpers/isDate');
const { getEvents, createEvent, updateEvent, deleteEvent } = require('../controllers/events');
const { validateFields } = require('../middlewares/validate-fields');
const { validateJWT } = require('../middlewares/validate-jwt');

const router = Router();

// Todas las rutas tienen que pasar por la validaci√≥n del JWT (middleware).
router.use(validateJWT);

// Obtener los eventos.
router.get('/', getEvents);

// Crear un nuevo evento.
router.post(
    '/',
    [
        check('title', 'Title is required.').notEmpty(),
        check('start', 'Start date is required').custom(isDate),
        check('end', 'End date is required').custom(isDate),
        validateFields
    ],
    createEvent
);

// Actualizar un evento.
router.put(
    '/:id',
    [
        check('title', 'Title is required.').notEmpty(),
        check('start', 'Start date is required').custom(isDate),
        check('end', 'End date is required').custom(isDate),
        validateFields
    ],
    updateEvent
);

// Borrar un evento
router.delete('/:id', deleteEvent);

module.exports = router;