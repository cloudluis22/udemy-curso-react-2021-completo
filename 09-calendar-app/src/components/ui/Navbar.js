import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { eventLogoutCleaning } from '../../actions/events';

export const Navbar = () => {

    const dispatch = useDispatch();

    const { name } = useSelector(state => state.auth)
    const handleLogout = () => {
        dispatch( startLogout() );
        dispatch( eventLogoutCleaning() );
    };


    return (
        <div className="navbar navbar-dark bg-dark mb-4">

            <span className="navbar-brand px-4"> { name } </span>

            <button className="btn btn-outline-danger mx-4 px-3" onClick={ handleLogout }>
                <i className="fas fa-sign-out-alt"></i>
                <span> Salir </span>
            </button>

        </div>
    )
}