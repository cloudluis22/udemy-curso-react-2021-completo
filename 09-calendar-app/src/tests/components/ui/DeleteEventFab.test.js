import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme from 'enzyme';

import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';

import '@testing-library/jest-dom';
import { eventStartDelete } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
    eventStartDelete: jest.fn()
}));

Enzyme.configure({ adapter: new Adapter() });

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);

store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <DeleteEventFab />
    </Provider>
);

describe('Pruebas en el componente de "<DeleteEventFab/>."', () => {

    test('Debe de mostrarse correctamente.', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('La funciÃ³n "eventStartDelete" debe de llamarse correctamente al hacer click.', () => {

        wrapper.find('button').simulate('click');
        expect(eventStartDelete).toHaveBeenCalled();

    })

});
