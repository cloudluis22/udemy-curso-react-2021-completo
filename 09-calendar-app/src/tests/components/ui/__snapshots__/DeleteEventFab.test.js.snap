// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente de "<DeleteEventFab/>." Debe de mostrarse correctamente. 1`] = `ReactWrapper {}`;
